buildtest/buildsystem/base.py

```diff
@@ -441,9 +418,13 @@ def run_tests(self, testfile):
 
         # Change to the test directory
         os.chdir(self._get_testdir())
```

Comment:
Is this does any ping to the file system to check for existence, it should be called once and then used. E.g.,
```python
testdir = self._get_testdir()
```
instead of being called multiple times to return the same folder.

Reply:
as of now (``devel`` HEAD) its only being called twice. See https://github.com/HPC-buildtest/buildtest-framework/blob/devel/buildtest/buildsystem/base.py 

It could be  useful to just retrieve via ``self.metadata.get("testdir")`` 

My only concern is I dont think we have os.path.expandvars set such that a test directory can be set using variable expansion.  This was done correctly in 

https://github.com/HPC-buildtest/buildtest-framework/blob/devel/buildtest/buildsystem/base.py#L268

We should probably have test to validate something like ``buildtest build --testdir $HOME`` and see if it writes test correctly.

Reply:
I think we should maybe aim to call this once, and store for the duration of the test? These sorts of setup variables we wouldn't want to weirdly change (for some buggy reason) in the middle of a test.

Reply:
i agree and if i understand the only reason why we are ``chdir`` before running test is to ensure any generated content from the test that may include compilation, executable, or output files get stored in test directory and not in current directory which would be ugly. Unrelated to this, i noticed that we don't remove test directory before writing test. This is an issue when some test fail upon re-execution since they assume a clean directory. 

This is true for example in ``examples/script/bzip2.yml``. The point is should test directory be clean to avoid rewriting content of generated file.

Reply:
I always get worried whenever scripts take charge to delete content - for example what if there was a bug with testdir and you shutil.rmtree something important? I think it makes sense to clean up the test directory on re-run, but is there some marker we can use to be absolutely sure that it's a test directory? Perhaps the output file or the folder structure? We haven't talked much about the structure of a test directory, but whatever that might encompass, this could be a good check to do.

Reply:
The other alternate is stick a build id in the sub-directory path to avoid rewriting to same directory, that's what i did before to avoid this issue and worked nice. Previously, buildtest would purge and write again, then this was avoided by added a unique sub-directory that is incremented. Every build was then written to ``build.json`` and there was a method to retrieve total build IDs so that buildtest knows what the next build ID is. 
Adding this complete feature would require a bit more refactoring and it could be its own PR. Right now everything is working as expected.

Another alternate to an ID is use a random hex (32bit) lets say to stick as build ID, later on buildtest would need to capture build ID in some json file to retrieve test reports. Again that's out of scope for another PR and feature request.

Reply:
I agree it's out of scope - I think it would be cleanest to remove the directory, but have some check that it's a build test directory, and maybe a user flag. I don't see it being useful to run multiple builds and have the folders build up (it was really messy last time) but if this approach is wanted, it would work to do a random hash each time (and make that hash very clear to the user in the screen logs). I think this is what I've seen other similar tools, either for workflows or test runs, do.

Reply:
I'll try to open a separate issue as this was related to Stampede test i was running. Some tests couldn't rerun after first time because directory existed. I think the value of reruns would be run test for reproducible results .

One good use-case is to run acceptance test once a week, and maybe performance test once a month. Right now all the tests for stampede2 are in one directory so i can just do ``buildtest build -c tests`` and it will run everything. Or i can go by sub-directory. But eventually i'd like to have this as repeated operation so buildtest can support multiple reruns cleanly so reports can be generated based on a build ID with timestamp on when it was run. 

I think the framework should support a way to gather data from test reports which users can understand.  Running ad-hoc commands once would not be useful unless all the data is results are captured somewhere that buildtest can query and report.